### PR TITLE
E2E tests for session handling

### PR DESCRIPTION
**Pull request by lukastaegert**
This extends the E2E test setup with tests for kicking users, taking over sessions and losing connectivity. Losing connectivity is simulated by instrumenting the WebSocket constructor and manually closing the connection or emitting an error.

It also contains some improvements I encountered while working on this:
- The local backend will now add a `statusCode: 410` to the error when a connection is closed so that connections can be cleaned up locally. I hope this works on AWS, i.e. the assumption about the 410 code is correct.
- When any other error is thrown while posting to a connection, it will never be rethrown but logged. This should ensure that messages are not swallowed for some clients.
- A user cannot take over their own session in the backend, i.e. when a user logs in and there already their connection is already registered for that name in the session, they will not kick themselves out immediately. While it should not be possible for this to happen in practice, I encountered it several times when using hot-module-replacement locally.
- When a connection item is not found, then an error is logged as this is always a bug that should not happen.
- The automatic reconnect would never reset its retry counter and did not have a time limit. This meant that when a user is on an unsteady connection, retry attempts will take ages after a while even though the user was connected before. This is improved in two ways:
  - Whenever a connection is established, the retry counter is reset
  - There is an upper limit of 4s for the retry interval
- When unmounting the Websocket, it will now be closed in a way so that it does not attempt to reconnect. This fixes issues when using hot module replacement.
- During development, only changes in the backend folder will now trigger the server to restart.

Furthermore, I added `role: alert` to the red text on the login window and a test-id to the reveal button to make it easy to find them in the tests.

### CODE REVIEW

File: backend/shared/database/getConnection.ts [hunk 1]
```diff
@@ -5,4 +5,10 @@ export const getConnection = ({
   connectionId,
   tableName,
   ddb,
-}: Config): Promise<ConnectionItem | void> => getItem('connectionId', connectionId, tableName, ddb);
+}: Config): Promise<ConnectionItem | undefined> => {
+  const connection = getItem<ConnectionItem>('connectionId', connectionId, tableName, ddb);
+  if (!connection) {
+    console.error('No connection found for connectionId', connectionId, connection);
```
Comment (ThiloAschebrock) [inline backend/shared/database/getConnection.ts:11]: What is the point of logging `connection` when it is always undefined?
```suggestion
    console.error('No connection found for connectionId', connectionId);
```

File: frontend/src/Components/RevealButton/RevealButton.tsx [hunk 2]
```diff
@@ -32,7 +32,12 @@ export const RevealButton = connectToWebSocket(
     const { missing, voted } = getNumberOfVotes(votes);
 
     return (
-      <button class={classes.revealButton} onClick={revealVotes} disabled={!connected || !voted}>
+      <button
+        data-testid="reveal-button"
```
Comment (ThiloAschebrock) [inline frontend/src/Components/RevealButton/RevealButton.tsx:36]: Can we use a meaningful but static aria label, such as `Reveal votes`, instead?

File: frontend/src/Components/WebSocket/WebSocket.tsx [hunk 3]
```diff
@@ -22,6 +22,8 @@ const initialWebSocketState: WebSocketState = {
 };
 
 const initialLoginData: WebSocketLoginData = { user: '', session: '' };
+const baseRetryWait = 200;
```
Comment (ThiloAschebrock) [inline frontend/src/Components/WebSocket/WebSocket.tsx:25]: Do we want to use all caps consistently for such constants?
```suggestion
const BASE_RETRY_WAIT = 200;
```

File: e2e/pages/card-page.ts [hunk 4]
```diff
@@ -35,6 +34,13 @@ export class CardPage {
     );
     await expect(this.votes).toHaveCount(states.length);
   }
+
+  async kickUser(name: string) {
+    const kickButton = this.votes
+      .filter({ has: this.page.locator(`:has-text("${name}")`) })
+      .locator('td:nth-child(3) button');
```
Comment (ThiloAschebrock) [inline e2e/pages/card-page.ts:41]: The button has the title attribute `Kick ${name}`. Can we find the button by this title instead?

**Review by ThiloAschebrock — APPROVED** [review]
Very nice! I just had some minor comments.

File: frontend/src/Components/RevealButton/RevealButton.tsx (hunk 2 above)
Comment (lukastaegert) [reply frontend/src/Components/RevealButton/RevealButton.tsx:36]: Definitely. I actually did this in the next PR #112 but I will move the change to this PR instead

File: e2e/pages/card-page.ts (hunk 4 above)
Comment (lukastaegert) [reply e2e/pages/card-page.ts]: Great idea, done!

File: frontend/src/Components/WebSocket/WebSocket.tsx [hunk 5]
```diff
@@ -22,7 +22,7 @@ const initialWebSocketState: WebSocketState = {
 };
 
 const initialLoginData: WebSocketLoginData = { user: '', session: '' };
-const baseRetryWait = 200;
+const BASE_RETRY_WAIT = 200;
 const maxRetryWait = 4000;
```
Comment (ThiloAschebrock) [inline frontend/src/Components/WebSocket/WebSocket.tsx:26]:
```suggestion
const MAX_RETRY_WAIT = 4000;
```

**Review by ThiloAschebrock — APPROVED** [review]

File: backend/sendmessage/src/send-message-to-connection.ts [hunk 6]
```diff
@@ -16,13 +16,10 @@ export const sendMessageToConnection = async (
...
   } catch (e) {
-    if (!(e instanceof Error)) {
-      return;
-    }
-    if ((e as AWSError).statusCode === 410) {
+    if (e && (e as AWSError).statusCode === 410) {
       return removeConnection(config);
     } else {
-      throw e;
+      console.error(`Unexpected error when posting message to connection ${connectionId}:`, e);
```
Comment (andreashille) [inline backend/sendmessage/src/send-message-to-connection.ts:22]: maybe i dont got it but why are we not rethrowing here?

**Review by andreashille — APPROVED** [review]

File: backend/sendmessage/src/send-message-to-connection.ts (hunk 6 above)
Comment (lukastaegert) [reply backend/sendmessage/src/send-message-to-connection.ts:24]: I added a comment to explain the reasoning better

**Comment by lukastaegert** [comment]
I also fixed the script name in the readme and added a comment about the docker image.

**Pull request merged**